### PR TITLE
Allow slices for SynchronizedArray

### LINKED ISSUE #11572: SynchronizedArray should support slicing
Hello everyone,

while checking multiprocessing-code, I found a bug in the following example:
```python
from multiprocessing import Value, Array
from ctypes import c_wchar_p, c_int

a = Array(c_int, 10)
print(a[0:4])
a[:] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```

This is roughly equivalent to the example-code in https://docs.python.org/3/library/multiprocessing.html#sharing-state-between-processes, and as expected, the code works fine on my python 3.11 (and 3.12).

If i run mypy (version 1.9.0, but from my understanding this should be a bug here in typeshed), i get
```
test.py:5: error: Invalid index type "slice" for "SynchronizedArray[c_int]"; expected type "int"  [index]
test.py:6: error: Invalid index type "slice" for "SynchronizedArray[c_int]"; expected type "int"  [index]
test.py:6: error: Incompatible types in assignment (expression has type "list[int]", target has type "c_int")  [assignment]
Found 3 errors in 1 file (checked 1 source file)
```

It seems as slicing the Array as done in the documentation is not supported in typeshed, as `SynchronizedArray` expects only int-indices. Also the return value of such an index seems to be off, as this one should be `list[int]`, and setting the value should be possible as documented.

Confusing to me, https://github.com/python/typeshed/blob/main/stdlib/multiprocessing/sharedctypes.pyi#L96 looks correct for this case, but mypy seems to pick `__getitem__` instead of `__getslice__`?

### PR DESCRIPTION
**Pull request by prauscher**
Fixes #11572 

This PR allows to use the syntax described in the python documentation: https://docs.python.org/3/library/multiprocessing.html#sharing-state-between-processes

Specifically:
```python
a = Array(c_int, 5)
print(a[0:2])
```

### CODE REVIEW

File: stdlib/multiprocessing/sharedctypes.pyi [hunk 1]
```diff
@@ -91,10 +91,14 @@ class Synchronized(SynchronizedBase[_SimpleCData[_T]], Generic[_T]):
...
+    @overload
+    def __getitem__(self, i: slice) -> list[_CT]: ...
+    @overload
     def __getitem__(self, i: int) -> _CT: ...
+    @overload
+    def __setitem__(self, i: slice, value: Iterable[_CT]) -> None: ...
+    @overload
     def __setitem__(self, i: int, value: _CT) -> None: ...
-    def __getslice__(self, start: int, stop: int) -> list[_CT]: ...
-    def __setslice__(self, start: int, stop: int, values: Iterable[_CT]) -> None: ...
```
Comment (AlexWaygood) [inline stdlib/multiprocessing/sharedctypes.pyi:97]: Why did you delete these? They exist at runtime, even if they don't do anything special in Python 3. We shouldn't delete them from the stub if they exist at runtime

Comment (prauscher) [reply stdlib/multiprocessing/sharedctypes.pyi:97]: Sure, I read your message as they were deprecated so I thought I could clean up once I am there. Will re-add them

Comment (AlexWaygood) [reply stdlib/multiprocessing/sharedctypes.pyi]: We'd need to deprecate and remove them at runtime before we'd be able to remove them from the stub. The most important thing for the stub is to accurately reflect the reality at runtime. So even if a method really shouldn't exist anymore at runtime, we should still include it in the stub while it does

Comment (prauscher) [reply stdlib/multiprocessing/sharedctypes.pyi]: Ahh, okay. I went too fast and thought useless would be deprecated :D

**Comment by prauscher** [comment]
I do not want to be too demanding, but just to be sure I'm not missing anything: Is there anything left for me to help going forward to merge?

**Review by srittau — APPROVED** [review]
Thanks!

**Pull request merged**